source/extensions/path/uri_template_lib/uri_template.h

```diff
@@ -0,0 +1,66 @@
+#pragma once
+
+#include <string>
+
+#include "source/extensions/path/uri_template_lib/proto/uri_template_rewrite_segments.pb.h"
+#include "source/extensions/path/uri_template_lib/uri_template_internal.h"
+
+#include "absl/status/statusor.h"
+#include "absl/strings/string_view.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace UriTemplate {
+
+enum class RewriteStringKind { Variable, Literal };
+
+struct RewritePatternSegment {
+  RewritePatternSegment(absl::string_view value, RewriteStringKind kind)
+      : value_(value), kind_(kind) {}
+  absl::string_view value_;
+  RewriteStringKind kind_;
+};
+
+// Returns the regex pattern that is equivalent to the given url_pattern.
+// Used in the config pipeline to translate user given url pattern to
+// the safe regex Envoy can understand. Strips away any variable captures.
+absl::StatusOr<std::string> convertURLPatternSyntaxToRegex(absl::string_view url_pattern);
+
+// Parses the specified pattern into a sequence of segments (which are
+// either literals or variable names).
+absl::StatusOr<std::vector<RewritePatternSegment>> parseRewritePattern(absl::string_view pattern);
+
+// Returns the parsed Url rewrite pattern and processes variables to be used by
+// RewriteURLTemplatePattern() and in rewrite.
+// |capture_regex| should be the regex generated by ConvertURLPatternSyntaxToRegex().
+absl::StatusOr<envoy::extensions::uri_template::UriTemplateRewriteSegments>
+parseRewritePattern(absl::string_view pattern, absl::string_view capture_regex);
```

Comment:
I think we need the regex but I'm not 100% on the reason why. I didn't write the core of this and I'm hesitant to start rewriting large parts as there was a lot of thought but into the original design.

Reply:
Is the regex used in that function beyond the call to get the capture groups? If so, I don't see it. The only change I'm proposing is the data type of this function parameter, and calling NamedCapturingGroups() in the caller instead of the method. Is this a large rewrite?

Reply:
Can you explain what you mean by large or functional chance? It's literally calling `NamedCapturingGroups()` in one place instead of another. 

In any case, since we've dramatically chanced this PR since it was started I don't think that drifting from the internal implementation is problematic.

Reply:
I don't really understand the proposed change. Could you explain more?